### ISSUE TITLE
Unexpected error with function argument type

### ISSUE THREAD

**Issue by bijij** [microsoft/pyright#2907]
**Describe the bug**
I'm not sure where to start here, pyright is complaining that an argument cannot be passed to a callable with an unbound typevar in a specific situation.

**To Reproduce**
Given the code below pyright outputs the following.
```
test.py
    test.py:16:21 - error: Argument of type "Literal[False]" cannot be assigned to parameter "a" of type "T@foo" in function "foo"
    Type "Literal[False]" cannot be assigned to type "None" (reportGeneralTypeIssues)
 ```
 If the call to `foo` at the start has its argument replaced with a superclass of bool, or the `if a is ...` checks against anything other than a boolean this type-checks.

**Expected behavior**
This should just type-check

**Screenshots or Code**
```python
from typing import TypeVar, Union

T = TypeVar('T')

def foo(a: T) -> Union[bool, T]:
    ...


def main():

    a = foo(None)

    if a is True:
        return

    a or foo(False)

```

**VS Code extension or command-line**
pyright 1.1.212

**Comment by erictraut** [comment]
Yeah, I can see why that would be confusing. Here's what's happening.

Pyright uses a technique called [bidirectional type inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types) in a variety of cases to help inform the inferred type of an expression.

As a simple example, let's consider the following:
```python
def func(a: list[float]):
    x = a or [1, 2, 3]
```

Normally, the inferred type of the expression `[1, 2, 3]` would be `list[int]`, so the type of `a or [1, 2, 3]` would evaluate to `list[float] | list[int]`. But that's probably not what was intended. Instead, `[1, 2, 3]` should probably be interpreted as type `list[float]` in this case. So pyright uses bidirectional type inference for the `or` operator to "nudge" the type of the right-hand-side operand to be consistent with the left-hand-side operand if possible.

Bidirectional type inference is also applied for call expressions. Consider the following:
```python
def foo(a: list[T]) -> list[T]: ...

x: list[float] = foo([1, 2, 3])
```

Like above, the inferred type of expression `[1, 2, 3]` would normally be `list[int]`. But in this case, that would lead to a type violation because `foo([1, 2, 3])` would also have a type of `list[int]` which isn't compatible with `list[float]`. 

In your code sample, bidirectional type inference is being applied for both the `or` operator and the call expression `foo(False)`. The type of `a` has been narrowed to `Literal[False] | None` based on the conditional statement, so pyright attempts to "nudge" the type of the right-hand expression (`foo(False)`) to be `Literal[False] | None` as well. For this to evaluate to `Literal[False] | None`, `T` must be `None`.

This approach of using bidirectional type inference is normally a very good thing because it leads to fewer false positives and eliminates the need for many explicit type annotations. But in this particular case, it's producing an undesirable behavior.

Part of the challenge here is the `Union[bool, T]`. Any time you create a union that contains a TypeVar like `T`, you're creating ambiguities for the TypeVar constraint solver.

I'll need to think more about whether I can add some heuristics to improve the behavior in this case.

In the meantime, I can offer you several workarounds:
1. Replace `a or foo(False)` with `foo(False)` since you know that `a` is either `False` or `None` at that point
2. Replace `a or foo(False)` with `a if a else foo(False)`
3. Assign the value of `foo(False)` to a temporary variable (as in `x = foo(False)` followed by `a or x`).

**Comment by erictraut** [comment]
I found a relatively clean way to solve the problem. It has a bit of a performance impact, but it's not too bad.

This fix will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.213, which I just published. It will also be included in the next release of pylance.
